### PR TITLE
writer-json: do not needlessly duplicate messages in SARIF

### PR DESCRIPTION
**Pull request by lzaoral**
... when we have only one.

### CODE REVIEW

File: tests/csgrep/90-sarif-writer-illegal-utf8-sequence-stdout.txt [hunk 1]
```diff
@@ -25,34 +25,7 @@
...
-                                                    }
-                                                },
-                                                "message": {
-                                                    "text": "illegal sequence '�' will be replaced with '�'"
-                                                }
-                                            },
-                                            "nestingLevel": 0,
-                                            "kinds": [
-                                                "warning"
-                                            ]
```
Comment (lzaoral) [inline tests/csgrep/90-sarif-writer-illegal-utf8-sequence-stdout.txt:49]: Note that we will lose also these nodes.  However, they are not documented in GitHub's SARIF manual [1] so I guess it's harmless?

[1] https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#location-object

Comment (kdudka) [reply tests/csgrep/90-sarif-writer-illegal-utf8-sequence-stdout.txt:49]: This is fine as long as the output of `csgrep --mode=sarif | csgrep --mode=json` stays unchanged.  For defects with more than one event, the `kinds.warning` field is important for csgrep to preserve data.  But such defects should be unaffected by this change.

File: src/json-writer.cc [hunk 2]
```diff
@@ -302,33 +302,35 @@ void SarifTreeEncoder::appendDef(const Defect &def)
     sarifEncodeMsg(&result, keyEvt.msg);
 
     // other events
-    PTree flowLocs, relatedLocs;
-    for (unsigned i = 0; i < def.events.size(); ++i) {
-        if (def.events[i].event == "#")
-            sarifEncodeComment(&relatedLocs, def, i);
-        else
-            sarifEncodeEvt(&flowLocs, def, i);
+    if (def.events.size() > 1) {
```
Comment (kdudka) [inline src/json-writer.cc:305]: Please write this as `1U < def.events.size()`.  The source code of csdiff always puts lower operand to left for consistency.

**Review by kdudka — APPROVED** [review]

**Review by jamacku — COMMENTED** [review]
Hmm, it looks like GitHub doesn't care if "`codeFlows`" message is missing. It duplicates it anyway.

Screenshot from 2022-08-23 14-06-01

<details>
<summary>But SARIF generated by csdiff looks correctly </summary>

```json
{
    "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
    "version": "2.1.0",
    "runs": [
        {
            "tool": {
                "driver": {
                    "name": "ShellCheck",
                    "version": "2.6.0.20220823.123456.g064f632.pr_77",
                    "informationUri": "https://github.com/csutils/csdiff"
                }
            },
            "results": [
                {
                    "ruleId": "SHELLCHECK_WARNING: note[SC2086]",
                    "locations": [
                        {
                            "id": 0,
                            "physicalLocation": {
                                "artifactLocation": {
                                    "uri": "src/index.sh"
                                },
                                "region": {
                                    "startLine": 9,
                                    "startColumn": 6
                                }
                            }
                        }
                    ],
                    "message": {
                        "text": "Double quote to prevent globbing and word splitting."
                    }
                },
                {
                    "ruleId": "SHELLCHECK_WARNING: warning[SC1090]",
                    "locations": [
                        {
                            "id": 0,
                            "physicalLocation": {
                                "artifactLocation": {
                                    "uri": "src/index.sh"
                                },
                                "region": {
                                    "startLine": 11,
                                    "startColumn": 3
                                }
                            }
                        }
                    ],
                    "message": {
                        "text": "ShellCheck can't follow non-constant source. Use a directive to specify location."
                    }
                },
                {
                    "ruleId": "SHELLCHECK_WARNING: note[SC2086]",
                    "locations": [
                        {
                            "id": 0,
                            "physicalLocation": {
                                "artifactLocation": {
                                    "uri": "src/index.sh"
                                },
                                "region": {
                                    "startLine": 11,
                                    "startColumn": 3
                                }
                            }
                        }
                    ],
                    "message": {
                        "text": "Double quote to prevent globbing and word splitting."
                    }
                },
                {
                    "ruleId": "SHELLCHECK_WARNING: note[SC2250]",
                    "locations": [
                        {
                            "id": 0,
                            "physicalLocation": {
                                "artifactLocation": {
                                    "uri": "src/index.sh"
                                },
                                "region": {
                                    "startLine": 120,
                                    "startColumn": 13
                                }
                            }
                        }
                    ],
                    "message": {
                        "text": "Prefer putting braces around variable references even when not strictly required."
                    }
                },
                {
                    "ruleId": "SHELLCHECK_WARNING: note[SC2250]",
                    "locations": [
                        {
                            "id": 0,
                            "physicalLocation": {
                                "artifactLocation": {
                                    "uri": "src/index.sh"
                                },
                                "region": {
                                    "startLine": 122,
                                    "startColumn": 6
                                }
                            }
                        }
                    ],
                    "message": {
                        "text": "Prefer putting braces around variable references even when not strictly required."
                    }
                }
            ]
        }
    ]
}
```

</details>

- https://github.com/jamacku/differential-shellcheck/pull/28

**Review by jamacku — APPROVED** [review]
LGTM

**Comment by lzaoral** [comment]
I'll mark this as a WIP since the messages are still duplicated in the visualised output.

**Comment by lzaoral** [comment]
@jamacku, aren't the duplicities actually absent with #68?  Maybe the second line is always sourced from the rule metadata and duplicated only when it's missing.

**Comment by lzaoral** [comment]
That theory is debunked by https://github.com/redhat-plumbers-in-action/differential-shellcheck/pull/117, so the search continues...

**Comment by lzaoral** [comment]
Hmm, this PR seems to be stuck since it is out of date with actual contents of https://github.com/lzaoral/csdiff/tree/dont-duplicate-messages and I've already tried pushing few times.

**Comment by lzaoral** [comment]
@jamacku Could you please create a issue to track the problem depicted in https://github.com/csutils/csdiff/pull/77#pullrequestreview-1082040907, if it is still present?  Thank you!

**Comment by lzaoral** [comment]
After discussion with @jamacku and @kdudka, we have decided to close this PR because it does not introduce any visible improvements for Differential ShellCheck.

**Pull request closed without merging**